src/network_interfaces.jl

```diff
@@ -0,0 +1,136 @@
+end
+
+function _next(r::Base.RefValue{Ptr{uv_interface_address_t}})
+    next_addr = r[] + sizeof_uv_interface_address_t
+    Ref(Ptr{uv_interface_address_t}(next_addr))
+end
+
+_is_loopback(addr) = 1 == @ccall jl_uv_interface_address_is_internal(addr::Ptr{uv_interface_address_t})::Cint
+
+_sockaddr(addr) = @ccall jl_uv_interface_address_sockaddr(addr::Ptr{uv_interface_address_t})::Ptr{Cvoid}
+
+_sockaddr_is_ip4(sockaddr::Ptr{Cvoid}) = 1 == @ccall jl_sockaddr_is_ip4(sockaddr::Ptr{Cvoid})::Cint
+
+_sockaddr_is_ip6(sockaddr::Ptr{Cvoid}) = 1 == @ccall jl_sockaddr_is_ip6(sockaddr::Ptr{Cvoid})::Cint
+
+_sockaddr_to_ip4(sockaddr::Ptr{Cvoid}) = IPv4(ntoh(@ccall jl_sockaddr_host4(sockaddr::Ptr{Cvoid})::Cuint))
+
+function _sockaddr_to_ip6(sockaddr::Ptr{Cvoid})
+    addr6 = Ref{UInt128}()
+    @ccall jl_sockaddr_host6(sockaddr::Ptr{Cvoid}, addr6::Ptr{UInt128})::Cuint
+    IPv6(ntoh(addr6[]))
+end
+
+# Define a selection of hardware types that we're interested in. Values taken from:
+# https://github.com/torvalds/linux/blob/28eb75e178d389d325f1666e422bc13bbbb9804c/include/uapi/linux/if_arp.h#L29
+@enum ARPHardware begin
+    ARPHardware_Ethernet = 1
+    ARPHardware_Infiniband = 32
+    ARPHardware_Loopback = 772
+end
+
+struct Interface
+    name::String
+    version::Symbol
+    ip::IPAddr
+
+    # These two fields are taken from the sysfs /type and /speed files if available:
+    # https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-class-net
+    type::Union{ARPHardware, Nothing}
+    speed::Union{Float64, Nothing}
+end
+
+function _get_interfaces(
+    ::Type{T}=IPAddr; loopback::Bool=false
+) where T <: IPAddr
+    addr_ref  = Ref{Ptr{uv_interface_address_t}}(C_NULL)
+    count_ref = Ref{Int32}(1)
+
+    err = uv_interface_addresses(addr_ref, count_ref)
+    if err != 0
+        error("Call to uv_interface_addresses() to list network interfaces failed: $(err)")
+    end
+
+    interface_data = Interface[]
+    current_addr = addr_ref
+    for i = 0:(count_ref[]-1)
+        # Skip loopback devices, if so required
+        if (!loopback) && _is_loopback(current_addr[])
+            # Don't don't forget to iterate the address pointer though!
+            current_addr = _next(current_addr)
+            continue
+        end
+
+        # Interface name string. The name is the first field of the struct so we
+        # just cast the struct pointer to a Ptr{Cstring} and load it.
+        name_ptr = unsafe_load(Ptr{Cstring}(current_addr[]))
+        name = unsafe_string(name_ptr)
+
+        # Sockaddr used to load IPv4, or IPv6 addresses
+        sockaddr = _sockaddr(current_addr[])
+
+        # Load IP addresses
+        (ip_type, ip_address) = if IPv4 <: T && _sockaddr_is_ip4(sockaddr)
+            (:v4, _sockaddr_to_ip4(sockaddr))
+        elseif IPv6 <: T && _sockaddr_is_ip6(sockaddr)
+            (:v6, _sockaddr_to_ip6(sockaddr))
+        else
+            (:skip, nothing)
+        end
+
+        # Load sysfs info
```

Comment:
We should gate this with `@static if Sys.isunix()` so that we don't do these file checks on Windows (we know they'll always fail, but it costs a few extra syscalls).

Reply:
Fixed in f4e0443.